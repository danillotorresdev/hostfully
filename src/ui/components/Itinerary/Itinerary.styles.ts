import { styled } from "@linaria/react";

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding: 40px 0;
  width: 100%;
  max-width: 600px;
  position: relative;
`;

export const CardHeader = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 10px;
`;

export const CardCTAs = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 10px;
  position: absolute;
  right: 0px;
  bottom: 24px;
  button {
    margin: 0 10px;
  }
  @media (min-width: 768px) {
    top: 0;
    bottom: auto;
  }
`;

export const Title = styled.span`
  font-size: 1.5rem;
  line-height: 133%;
  font-weight: 700;
  color: #2d2aa5;
`;

export const Subtitle = styled.span`
  font-size: 1rem;
  display: flex;
  span {
    display: flex;
    flex-direction: column;
    font-weight: 700;
    margin: 0 5px;
  }
  font-size: 0.875rem;
`;

export const DeleteButton = styled.button``;

export const CardBody = styled.div`
  display: flex;
  flex-direction: column;
  padding-bottom: 20px;
`;

export const CardFooter = styled.div`
  font-size: 0.875rem;
  margin-top: 10px;
  color: rgb(0, 0, 20);
  display: flex;
  align-items: center;
  font-weight: 700;
`;
