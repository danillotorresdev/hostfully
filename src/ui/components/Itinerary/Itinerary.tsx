import * as S from "./Itinerary.styles";
import { Button } from "@/ui/components/Button/Button";
import { Icon } from "@/ui/components/Icon/Icon";
import { formatPassengers } from "@/utils/functions/formatPassengers/formatPassengers";
import { formatDateToLocaleString } from "@/utils/functions/formatDateToLocaleString/formatDateToLocaleString";
import { IntinetaryRoute } from "@/ui/components/Itinerary/IntinetaryRoute/IntinetaryRoute";
import { EditBookingModal } from "@/ui/components/Modals/EditBookingModal/EditBookingModal";
import { useState } from "react";

export type Booking = {
  id: string;
  origin: string;
  destination: string;
  startDate: string;
  endDate: string;
  passengers: {
    adults: number;
    children: number;
  };
};

type BookingItemProps = {
  booking: Booking;
  deleteBooking: (id: string) => void;
};

export const Itinerary = ({ booking, deleteBooking }: BookingItemProps) => {
  const [showModal, setShowModal] = useState(false);

  const handleShowModal = () => {
    setShowModal(!showModal);
  };

  return (
    <S.Container>
      <S.CardHeader>
        <S.Title>{booking.destination}</S.Title>
        <S.Subtitle>
          Booking for {formatPassengers(booking.passengers)}
          {`  | ${formatDateToLocaleString({
            date: booking.startDate,
            options: { month: "short", day: "numeric", year: "numeric" },
          })} to ${formatDateToLocaleString({
            date: booking.endDate,
            options: { month: "short", day: "numeric", year: "numeric" },
          })}`}
        </S.Subtitle>

        <S.CardCTAs>
          <Button
            onClick={handleShowModal}
            size="md"
            noBackground
            noPadding
            title="Edit booking"
          >
            <Icon name="edit" size="md" color="rgb(77 75 75)" />
          </Button>
          <Button
            onClick={() => deleteBooking(booking.id)}
            size="md"
            noBackground
            noPadding
            title="Delete booking"
          >
            <Icon name="trash" size="md" color="rgb(163 83 83)" />
          </Button>
        </S.CardCTAs>
      </S.CardHeader>
      <S.CardBody>
        <IntinetaryRoute booking={booking} type="DEPARTURE" />
        <IntinetaryRoute booking={booking} type="RETURN" />
      </S.CardBody>
      <S.CardFooter>
        <Icon name="members" size="sm" />
        {formatPassengers(booking.passengers)}
      </S.CardFooter>

      <EditBookingModal
        bookingId={booking.id}
        showModal={showModal}
        handleShowModal={handleShowModal}
      />
    </S.Container>
  );
};
