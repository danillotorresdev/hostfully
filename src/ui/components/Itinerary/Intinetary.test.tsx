import { Itinerary } from './Itinerary';
import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@/utils/settings/tests/utilities';

describe('Itinerary component', () => {
  const mockBooking = {
    id: '123',
    origin: 'New York',
    destination: 'Los Angeles',
    startDate: '2024-03-01',
    endDate: '2024-03-05',
    passengers: {
      adults: 2,
      children: 1,
    },
  };

  it('renders booking details correctly', () => {
  const {asFragment} =  render(<Itinerary booking={mockBooking} deleteBooking={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByText('Los Angeles')[0]).toBeDefined();
    expect(screen.getByText('Booking for 2 Adults and 1 Child | Feb 29, 2024 to Mar 4, 2024')).toBeDefined();
    expect(screen.getAllByText('2 Adults and 1 Child')[0]).toBeDefined();
  });

  it('displays edit modal when edit button is clicked', () => {
    render(<Itinerary booking={mockBooking} deleteBooking={() => {}} />);

    fireEvent.click(screen.getByTitle('Edit booking'));

    expect(screen.getByText('Edit Booking')).toBeDefined();
  });

  it('calls deleteBooking function when delete button is clicked', () => {
    const deleteBookingMock = vi.fn();
    render(<Itinerary booking={mockBooking} deleteBooking={deleteBookingMock} />);

    fireEvent.click(screen.getByTitle('Delete booking'));

    expect(deleteBookingMock).toHaveBeenCalledWith('123');
  });
});
