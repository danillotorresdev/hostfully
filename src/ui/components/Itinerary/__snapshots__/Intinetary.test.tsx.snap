// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Itinerary component > renders booking details correctly 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-0 csr9ikx"
  >
    <div
      class="mocked-styled-1 cx2fecm"
    >
      <span
        class="mocked-styled-3 ta36wq1"
      >
        Los Angeles
      </span>
      <span
        class="mocked-styled-4 s1a79fbf"
      >
        Booking for 2 Adults and 1 Child  | Feb 29, 2024 to Mar 4, 2024
      </span>
      <div
        class="mocked-styled-2 c16asigh"
      >
        <button
          class="mocked-styled-8 bb8cb6j"
          style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
          title="Edit booking"
        >
          <svg
            class="svg svg-edit"
            data-testid="icon-svg"
            style="width: 24px; height: 24px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#edit"
            />
          </svg>
        </button>
        <button
          class="mocked-styled-8 bb8cb6j"
          style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
          title="Delete booking"
        >
          <svg
            class="svg svg-trash"
            data-testid="icon-svg"
            style="width: 24px; height: 24px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#trash"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="mocked-styled-6 c6r2qql"
    >
      <div
        class="mocked-styled-9 ix3zkq7"
      >
        <div
          class="mocked-styled-10 c1n5q6aa"
        >
          <div>
            <svg
              class="svg svg-departure"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#departure"
              />
            </svg>
             
            <b>
              Departure
            </b>
          </div>
          <span
            class="mocked-styled-11 dsggpze"
          >
            Thu, Feb 29, 2024
          </span>
        </div>
        <div
          class="mocked-styled-12 c1yloby"
        >
          <span
            title="New York"
          >
            New York 
          </span>
          <svg
            class="svg svg-arrow-foward"
            data-testid="icon-svg"
            style="width: 16px; height: 16px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#arrow-foward"
            />
          </svg>
           
          <span
            title="Los Angeles"
          >
            Los Angeles
          </span>
        </div>
      </div>
      <div
        class="mocked-styled-9 ix3zkq7"
      >
        <div
          class="mocked-styled-10 c1n5q6aa"
        >
          <div>
            <svg
              class="svg svg-arrival"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#arrival"
              />
            </svg>
             
            <b>
              Return
            </b>
          </div>
          <span
            class="mocked-styled-11 dsggpze"
          >
            Mon, Mar 4, 2024
          </span>
        </div>
        <div
          class="mocked-styled-12 c1yloby"
        >
          <span
            title="New York"
          >
            New York 
          </span>
          <svg
            class="svg svg-arrow-back"
            data-testid="icon-svg"
            style="width: 16px; height: 16px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#arrow-back"
            />
          </svg>
           
          <span
            title="Los Angeles"
          >
            Los Angeles
          </span>
        </div>
      </div>
    </div>
    <div
      class="mocked-styled-7 c1wkgrf8"
    >
      <svg
        class="svg svg-members"
        data-testid="icon-svg"
        style="width: 16px; height: 16px;"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="/assets/sprite.svg#members"
        />
      </svg>
      2 Adults and 1 Child
    </div>
  </div>
</DocumentFragment>
`;
